codelab/01_kpt-validate/tutorial.md

```diff
@@ -0,0 +1,321 @@
+
+
+<walkthrough-pin-section-icon></walkthrough-pin-section-icon>
+**Glossary**
+> [2] `kpt function` is a kubernetes resource with a `config.kubernetes.io/function` annotation. 
+Read more [here](https://googlecontainertools.github.io/kpt/concepts/functions/)
+
+
+## Verifying the validation 
+
+### "happy path" 
+
+Let's run `skaffold dev`. Now it will validate the resource according to the functions in 
+<walkthrough-editor-select-line filePath="guestbook-cl/skaffold.yaml" startLine="14" endLine="14" startCharacterOffset="7" endCharacterOffset="13">`fnPath`, 
+and *then* deploy the resource to the cluster.
+
+```bash
+skaffold dev
+```
+
+
+### "sad path"
+
+Since the validation passes silently, let's break the kubeval to prove the validation works! 
+
+* In <walkthrough-editor-open-file filePath="guestbook-cl/config/frontend/deployment.yaml">config/frontend/deployment.yaml</walkthrough-editor-open-file>, 
+    remove the <walkthrough-editor-select-line filePath="guestbook-cl/config/frontend/deployment.yaml" startLine="18" endLine="18" startCharacterOffset="0" endCharacterOffset="30">spec.template.spec.containers.image</walkthrough-editor-select-line> field (in line 19)
+   
+    *You can **either** manually delete the line **or** run the following command*
+
+    ```bash
+    sed -i '19d' ./config/frontend/deployment.yaml
+    ```
+
+*   Check the skaffold dev output.
+    
+    ```bash
+    skaffold dev
+    ```
+    You should expect the following warning about the missing field.
+    
+    ```terminal
+    The Deployment "frontend" is invalid: spec.template.spec.containers[0].image: Required value
+    ```
+
+*   You can add back the removed line by copying the following command and running it in the terminal.  
+
+    ```shell script
+    sed -i '18 a \        image: "frontend"'     ./config/frontend/deployment.yaml
+    ```
+    
+<walkthrough-notification-menu-icon></walkthrough-notification-menu-icon>
+**Tips**
+> You can find all the `kpt` validation functions from this [catalog](https://googlecontainertools.github.io/kpt/guides/consumer/function/catalog/validators/)
+> 
+> Or write your own versions. See [instructions](https://googlecontainertools.github.io/kpt/guides/consumer/function/).
+    
+    
+## Validating through a pipeline
+Time to complete: **About 5 minutes**
+
+Instead of using a single validation function, we are more interested in applying the 
+configurations through a series of validation functions, each of which checks a specific rule.
+
+To do so, you can put a list of `kpt` functions into a single file and kpt will apply the 
+resources through these functions based on the function orders in the single file. 
+Like the graph shows.
+
+![image](https://github.com/GoogleContainerTools/skaffold/tree/master/logo/skaffold.png)
```

Comment:
```suggestion
![image](https://raw.githubusercontent.com/GoogleContainerTools/skaffold/master/logo/kpt-pipeline.png)
```
I believe this should be the graph since it is not referenced anywhere, and ```raw.githubusercontent.com``` should  be used, otherwise the image won't load properly.

Reply:
👍